co2calculator/calculate.py

```diff
@@ -453,208 +464,107 @@ def calc_co2_ferry(
-    elif distance is not None and (start is not None or destination is not None):
-        warnings.warn(
-            "Both distance and start/stop location were provided. "
-            "Only distance will be used for emission calculation."
-        )
-        stops = None
-    elif start is None and destination is None and distance is not None:
-        stops = None
-    elif start is not None and destination is not None and distance is None:
-        # check if stops are provided in the right form
-        if transportation_mode == "car" and (
-            type(start) != str or type(destination) != str
-        ):
-            raise ValueError(
-                "Wrong data type for start and destination."
-                "Please provide a three letter IATA code for train stations."
-            )
-        elif transportation_mode != "car" and (
-            type(start) != dict or type(destination) != dict
-        ):
-            raise ValueError(
-                "Wrong data type for start and destination."
-                "Please provide a dictionary."
+) -> Tuple[float, float, str, str]:
+    """Interface to calculate business trips"""
+    # TODO: Validate that type switch for `transportation_mode` is not a
+    # breaking change
+
+    # All co2 emissions are calculated for given distance. Hence, let's check
+    # if distance was parsed. If not, calculate that first.
+    # NOTE: Imho (Jakob): this should be taken care of by the caller (backend)
+
+    if not distance:
+        if start is None or destination is None:
+            raise InputMissing(
+                "for location based calculations: start and destination must not be None"
             )
 
-        stops = [start, destination]
-    if transportation_mode == "car":
-        emissions, dist = calc_co2_car(
-            distance=distance,
-            stops=stops,
-            passengers=passengers,
-            size=size,
-            fuel_type=fuel_type,
+        start_loc = Stop(**start)
+        dest_loc = Stop(**destination)
+        stops = StopList.parse_obj([start_loc, dest_loc])
+        dist_request = DistanceRequest(
+            stops=stops, transportation_mode=transportation_mode
         )
-    elif transportation_mode == "bus":
-        emissions, dist = calc_co2_bus(
-            size=size,
-            fuel_type=fuel_type,
-            occupancy=occupancy,
-            vehicle_range="long-distance",
+        if (
+            dist_request.transportation_mode is TransportationMode.CAR
+            or dist_request.transportation_mode is TransportationMode.MOTORBIKE
+        ):
+            distance = trip_direct(dist_request)
+        if transportation_mode in [
+            TransportationMode.BUS,
+            TransportationMode.TRAIN,
+            TransportationMode.FERRY,
+            TransportationMode.PLANE,
+        ]:
+            distance = trip_detour(dist_request)
```

Comment:
Nice, this is so clean! Only remark: `distance` is unbound: If - for some reason - the transportation mode is not listed here, we do not assign/calculate a value for `distance`. Hence, check for completeness of the checks or add an `else` block where you raise an exception or use one of the two as default/fallback